attest/core/src/ias/verifier.rs

```diff
@@ -0,0 +1,1138 @@
+                }
+
+                // Otherwise, check if any of the pubkeys in the chain are a trust anchor
+                for cert in &mut signer_chain {
+                    for cacert in &mut trust_anchors {
+                        if cert.public_key_mut().write_public_der_vec()
+                            == cacert.public_key_mut().write_public_der_vec()
+                        {
+                            return Some(());
+                        }
+                    }
+                }
+                None
+            })
+            .ok_or(VerifierError::BadSignature)?;
+
+        let report_data = VerificationReportData::try_from(report)?;
+
+        for verifier in &self.verifiers {
+            if !verifier.verify(&report_data) {
+                return Err(VerifierError::Verification(report_data));
+            }
+        }
+
+        Ok(report_data)
+    }
+}
+
+/// An enumeration of possible report-data verifier
+#[derive(Debug, Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
+enum VerifyIasReportDataType {
+    /// A verifier that checks the nonce of the IAS report
+    Nonce(NonceVerifier),
+    /// A verifier that checks the MRENCLAVE value and advisory IDs of the
+    /// examined report
+    Enclave(MrEnclaveVerifier),
+    /// A verifier that checks the MRSIGNER value and advisory IDs of the
+    /// examined report
+    Signer(MrSignerVerifier),
+    /// A verifier that chains to a list of quote verifiers.
+    Quote(QuoteVerifier),
+    /// A verifier that checks the PSE manifest hash and result
+    Pse(PseVerifier),
+}
+
+impl Verify<VerificationReportData> for VerifyIasReportDataType {
+    fn verify(&self, report_data: &VerificationReportData) -> bool {
+        match self {
+            VerifyIasReportDataType::Nonce(v) => v.verify(report_data),
+            VerifyIasReportDataType::Enclave(v) => v.verify(report_data),
+            VerifyIasReportDataType::Signer(v) => v.verify(report_data),
+            VerifyIasReportDataType::Quote(v) => v.verify(report_data),
+            VerifyIasReportDataType::Pse(v) => v.verify(report_data),
+        }
+    }
+}
+
+/// A [`VerifyIasReportData`] implementation that will check report data for the
+/// presence of the given IAS nonce.
+#[derive(Debug, Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
+struct NonceVerifier {
+    /// The nonce to be checked for.
+    pub nonce: IasNonce,
+}
+
+impl Verify<VerificationReportData> for NonceVerifier {
+    fn verify(&self, data: &VerificationReportData) -> bool {
+        data.nonce
+            .as_ref()
+            .map(|v| v.eq(&self.nonce))
+            .unwrap_or(false)
+    }
+}
+
+/// A helper function used to check exceptions to the quote error = fail rule.
+fn check_ids(quote_status: &IasQuoteResult, config_ids: &[String], sw_ids: &[String]) -> bool {
+    match quote_status {
+        Ok(_) => true,
+        Err(IasQuoteError::ConfigurationNeeded { advisory_ids, .. }) => {
+            for id in advisory_ids {
```

Comment:
This looks like a fun place for iter().all(), sometihng like:

```
advisory_ids.iter().all(|id| config_ids.contains(id)
```

Just a suggestion, do whatever you prefer.